lib/pure/optionsutils.nim

```diff
@@ -0,0 +1,160 @@
+## dot-chaining but stops if the left hand side is a none-option.
+
+import options, macros
+
+proc map*[T](self: Option[T], callback: proc (input: T)) =
+  ## Applies a callback to the value in this Option
+  if self.isSome:
+    callback(self.unsafeGet)
+
+proc map*[T, R](self: Option[T], callback: proc (input: T): R): Option[R] =
+  ## Applies a callback to the value in this Option and returns an option
+  ## containing the new value. If this option is None, None will be returned
+  if self.isSome:
+    some[R]( callback(self.unsafeGet) )
+  else:
+    none(R)
+
+proc flatten*[A](self: Option[Option[A]]): Option[A] =
+  ## Remove one level of structure in a nested Option.
+  if self.isSome:
+    self.unsafeGet
+  else:
+    none(A)
+
+proc flatMap*[A, B](self: Option[A], callback: proc (input: A): Option[B]): Option[B] =
+  ## Applies a callback to the value in this Option and returns an
+  ## option containing the new value. If this option is None, None will be
+  ## returned. Similar to ``map``, with the difference that the callback
+  ## returns an Option, not a raw value. This allows multiple procs with a
+  ## signature of ``A -> Option[B]`` (including A = B) to be chained together.
+  map(self, callback).flatten()
+
+proc filter*[T](self: Option[T], callback: proc (input: T): bool): Option[T] =
+  ## Applies a callback to the value in this Option. If the callback returns
+  ## `true`, the option is returned as a Some. If it returns false, it is
+  ## returned as a None.
+  if self.isSome and not callback(self.unsafeGet):
+    none(T)
+  else:
+    self
+
+macro `?.`*(option: untyped, statements: untyped): untyped =
+  ## Existential operator. Works like regular dot-chaining, but if
+  ## the left had side is a ``none`` then the right hand side is not evaluated.
+  ## In the case that ``statements`` return something the return type of this
+  ## will be ``Option[T]`` where ``T`` is the returned type of ``statements``.
+  ## If nothing is returned from ``statements`` this returns nothing.
+  ##
+  ## .. code-block:: nim
+  ##   echo some("Hello")?.find('l') ## Prints out Some(2)
+  ##   some("Hello")?.find('l').echo # Prints out 2
+  ##   none(string)?.find('l').echo # Doesn't print out anything
+  ##   echo none(string)?.find('l') # Prints out None[int] (return type of find)
+  ##   # These also work in things like ifs
+  ##   if some("Hello")?.find('l') == 2:
+  ##     echo "This prints"
+  ##   if none(string)?.find('l') == 2:
+  ##     echo "This doesn't"
+  let opt = genSym(nskLet)
+  var
+    injected = statements
+    firstBarren = statements
+  if firstBarren.len != 0:
+    while true:
+      if firstBarren[0].len == 0:
+        firstBarren[0] = nnkDotExpr.newTree(
+          nnkDotExpr.newTree(opt, newIdentNode("unsafeGet")), firstBarren[0])
+        break
+      firstBarren = firstBarren[0]
+  else:
+    injected = nnkDotExpr.newTree(
+      nnkDotExpr.newTree(opt, newIdentNode("unsafeGet")), firstBarren)
+
+  result = quote do:
```

Comment:
this pattern (https://en.wikipedia.org/wiki/Immediately-invoked_function_expression) in Nim can use block expressions:
```
var i = 0
let a = block:
  if i == 0:
    i+2
  else:
    i*3
echo a
```

Reply:
The logic in the current procedure is not that simple though, this is what would be generated for something that returns a value:

```
(proc (): auto =
  let :tmp130615 = x
  if :tmp130615.isSome:
    return some(:tmp130615.unsafeGet.slice(3)[0]))()
```
As you can see it uses auto as the type and relies on Nim to create the default `none` value regardless of the type, so it doesn't have an else branch to the if case. I tried wrapping this up in a block statement, but without luck.

Reply:
```suggestion
  result = quote do:
    (proc (): auto {.inline.} =
```

Reply:
Yeah making it inline is probably a good idea, but if I apply your suggested cange it would be a double proc wouldn't it? Not sure how this new, fancy, GitHub feature works :P

Reply:
oh i see, guess I used that new feature wrong myself; maybe just apply the change manually then

Reply:
retried here: https://github.com/nim-lang/Nim/pull/9161/files/753e117925b74c85e133137bdd3d6fcef8ea501c#r225682964